### ISSUE TITLE
Declare winner of A/B test automatically

### ISSUE THREAD

**Issue by escopecz** [mautic/mautic#1752]
## Description

A/B test is a cool feature and a must for data based decision making, but right now, Mautic doesn't automate the winner. I suggest to let user define the end of the A/B test (for example after 5000 page hits) and then use only the winner. Currently, a user has to manually (periodically?) control the test and then unpublish the variants which have lost.

**Comment by ps92** [comment]
A lot of factors need to be considered for this feature. 
When should you actually decide the winners? Should you actually decide it after the unpublished date? 
What if the major transactions occur after the emails are unpublished and the game changes? Will that not mess up the emails already sent with the result of the former winner?

**Comment by escopecz** [comment]
Thanks for the questions, any discussion about this topic is welcome. 

> When should you actually decide the winners?

That's up to the marketer to decide. It can be after a certain amount of send emails or page hits. I don't thin a time range is a good metric since it might not provide sufficient statistical sample.

> Should you actually decide it after the unpublished date?

How is the unpublish date related to A/B tests?

> What if the major transactions occur after the emails are unpublished and the game changes?

Whenever a one of the A/B tested emails/pages changes, the test is restarted to that moment. That's how it works even now.

> Will that not mess up the emails already sent with the result of the former winner?

There is no way we can change already sent emails. Or maybe I don't understand the question. In that case please rephrase.

**Comment by ps92** [comment]
Yes, even I think that time is not a good metric. But number or percent of total hits should also not be considered, we need to come up with something better to actually decide the winners. 
For I was considering the time as a metric, I thought of using the unpublish date to be a classifier for time point when you can actually decide the winner(s). 

For transactions, I meant the number of opens or clicks. If at a point of time, variant A is decided as winner because of current stats on day 1, and maybe a day later (on day 2), variant B is opened by almost everyone, then on day 2, variant B is the winner. But if we already sent mails considering A as variant, then day 2 would be confusing I guess. 

I think all of my doubts is because I am considering time as a metric.

**Comment by PLIDS** [comment]
The A/B test feature would get much easier to implement if done through campaign builder. Mautic just needs two additional elements in the campaign builder:
1.Conditions element "Random"
2.Conditions element "Stats"
With these two elements, you can create any A/B test you want, including any decision
1.create the email variations X and Y you want to test
2.Split your contacts into two segments "Start"/"Final"
3.setup a 1st campaign using segments "Start" and using the "Random" conditions element to send the variations X and Y
4.move the contacts in case of success to different segments S-X and S-Y, depending on variation
5.create a 2nd campaign using the "Stats" conditions element that compares numbers of elements in S-X and S-Y and then sends the winning email variation to segment "Final"

So instead of creating a separate huge A/B-feature with lots of implementation and testing effort, just adding two elements to the campaign builder gives Mautic same possibilities, but many more options with less effort

**Comment by escopecz** [comment]
Thanks for the ideas!

 Actually, the A/B tests for emails and pages are currently available in Mautic. All this feature request is about is to automatically stop the A/B test after some statistically significant sample of sent emails/page hits, declare the winner and use that winner for the rest. What you suggest is more complicated. Or at least as I understand it.

**Comment by PLIDS** [comment]
Yes, we have used the A/B tests for emails and pages already. But we were already wondering why this separate feature exists at all and is not integrated into the campaign workflow. Because if the campaign workflow would have a simple conditions element "Random" with probability parameter, it would already now work in the same way as the separate A/B feature. Now, with this feature request, the A/B test is even more turning into a workflow, because now you want to add a condition and a following sequence. So we were wondering even more why this is not integrated into the campaign workflow builder.

**Comment by escopecz** [comment]
> Now, with this feature request, the A/B test is even more turning into a workflow, because now you want to add a condition and a following sequence.

It actually doesn't have anything to do with a workflow. Or at least as I see it, you'd need one more field to add to the A/B tested email/page root to define when the test should end. That's it. No need to create separate campaign conditions or 2 different campaigns to build a A/B test. Your idea also wouldn't work for landing pages.

**Comment by PLIDS** [comment]
Right for the landing page it would require one additional field in campaign builder "Send Emails" Actions element that specifies the Landing page of the Email Form. But that is probably not a huge effort.
Please dont get us wrong: We like the campaign builder feature so much, that we hope that development is not loosing focus on extending its abilities - that is all :-)

**Comment by devfss** [comment]
I agree with your goal but I only partially agree with what's been suggested so far.  I think this should be configurable to allow for various marketing situations and give the marketer the control and flexibility to decide.  I haven't fully thought this through, so this could just be me thinking out loud.

**A/B Testing should allow for various constraints and actions** determined by those constraints.  So we'd have the variable that we're testing and then constraints around any control variables.

> automatically stop the A/B test after some statistically significant sample

This is just a possible test constraint.  We could have constraints to end the test or constraints to interrupt or pause the test or send an alert.  (example: a web page with fairly consistent traffic, if we setup the A/B test to test click through rate [test variable] and a constraint setup based on average plus or minus two standard deviations of average daily visitors.  Then we let the test start and after a few days suddenly 0 traffic for a given time period we now have a situation outside the constraint parameters which triggers: an alert, not counting that outlying data in the results, or an invalidation of the whole test.)

> I was considering the time as a metric

I think time can be both a reporting metric as well as constraint.  Sometimes a marketing campaign is time sensitive (not the best use case for A/B testing but some people might still need to try) so options for ending the A/B Test based on time should still be valid and possible.  The best solution should allow for various testing constraints and should be fully configurable based on the test.

I'd like to be able to see and be able to act on greater reporting analytics based on Time ranges that emails are opened or A/B Testing to see the influence Time has on opens (example: sending staggered email 2 days apart vs 3 days apart etc)

**So our requested feature could be changed to: give Constraint Options and for greater usability even show which option is considered best practice by being marked as recommended.**

**Example choices for constraint to end the test:**
- Computed Statistical significant sample based on using a ratio of total to determine sample size (assumes total is known) _[recommended]_
- Sliding statistical sample (similar to Google's method of giving priority to the side that is pulling ahead)
- Predetermined sample size defined by marketer
- Predetermined time limit defined by marketer (even if sample size is low then a "win" based on a couple of percent is still just as good as if the marketer had taken a best guess for the whole campaign)
- some method someone extended in the future

I think this will be both flexible and allow for it to be extended while promoting best practices.

**Comment by massudaw** [comment]
I have a proposal and i am willing to implement this feature. 
My proposal covers the campaign builder. I propose adding a new triggerMode `'abtest'` to perform AB testing. 
 This new trigger mode is going to have some new parameters
  -  `winner_rollout_date` or `winner_rollout_interval`   for configuring how much time the test will last
  -  `sample_size` for configuring how many leads participate in the sample group

Processing:
Then with this information the processing is done all inside the campaign processing.
The new trigger mode is processed in two steps . Both of them are mapped to scheduled events. 
The first one is the sample group and they receive randomized email and is scheduled to start on the `triggerDate`.
The second one is the main group and is scheduled to star on the `winner_rollout_date`  or `winner_rollout_interval` after the `triggerDate`. Before processing this group we check if the rollout date  has passed and  pick the winner based on the abtest criteria. Then  we execute the triggerScheduledEvents loop.

Any comments will be appreciated.

**Comment by PLIDS** [comment]
@massudaw +1 Great approach. That is exactly what we thought about.

**Comment by escopecz** [comment]
@massudaw Love it! Should I mark this issue as Work In Progress so other developers would know it's in your hands?

**Comment by TomRoethlisberger** [comment]
@massudaw With that approach, you can take out the time constraint from the discussion. I have also thought about an approach for the automated decision. In special, I'm interested at the decision taking for the winner. To keep it simple, I would propose the following:

- Define two criteria for the marketeer to choose: Opening rate and click through rate of one link within the email
- the sampling size is not that important as long as it is significant. I would propose at least 100 emails. That would allow us to use A/B testing also for small clients
 - with that sample size, we should use a Bayesian approach for the decision making. With that we don't need a large sample size since we don't need statistical significance

Let me know if we can help here. I'm eager to create more intelligence with my team :-)

**Comment by TomRoethlisberger** [comment]
@escopecz Did you hear from @massudaw for the implementation of the feature? If not, I would like to offer our services to implement the feature. If you agree you could mark it as work in progress.

**Comment by escopecz** [comment]
I let @massudaw to speak for himself. If he doesn't answer soon, consider this issue to be free to be worked on. I'll mark it as WIP either way.

**Comment by dbhurley** [comment]
Any updates on this particular feature? @massudaw or @TomRoethlisberger ?

**Comment by AkkuFresh** [comment]
Will this feature available in the self hosted Mautic instances? It would be very useful tool with integration in the Campaign builder as it was described earlier.

**Comment by kuzmany** [comment]
@massudaw or @TomRoethlisberger any updates on this feature?

**Comment by AlbanL74fr** [comment]
About that

"Whenever a one of the A/B tested emails/pages changes, the test is restarted to that moment. That's how it works even now."

I believe it should prompt the user, at save to ask whether or not we want to the test to start all over again.

**Comment by escopecz** [comment]
@ondrejsibl is working on this. Should be PRed soon. I'll return the WIP back.

**Comment by RCheesley** [comment]
@ondrejsibl did this get over the line at all, or is it still WIP?

**Comment by RCheesley** [comment]
@escopecz I believe this is in use on Mautic Cloud - is it going to be PR'd to the community?

**Comment by escopecz** [comment]
Yes

**Comment by thechrise** [comment]
@escopecz any word on this PR? Our company is looking to make the switch, and this would certainly help.

**Comment by escopecz** [comment]
We have this feature ready on our fork. But we cannot even create a PR for community because there are so many changes already waiting to be merged. The A/B winner code is built on those changes. The community product team is working on getting the open PRs merged. The Acquia team will help. We will create the PR as soon as we'll be able to.

**Comment by thechrise** [comment]
Thanks for your quick response, @escopecz.

**Comment by RCheesley** [comment]
@escopecz wondering what PR's are blocking this coming over to the community? Let's aim to get it unblocked for 3.1?

**Comment by RCheesley** [comment]
@escopecz is it likely that we can get this over for the 3.3 release?

**Comment by escopecz** [comment]
@RCheesley I was half-way through cherry-picking the 70-commit PR with automatic A/B test winner but found a blocking PR https://github.com/acquia/mc-cs/pull/324 (internal) that used to be blocked by https://github.com/mautic/mautic/pull/7094 but it's not anymore. So I'll ask the author to push it to the community now.

**Comment by RCheesley** [comment]
Awesome, thank you for following up on that! Hopefully we can now start to move through these PR's at a decent pace and get things moving!

**Comment by kuzmany** [comment]
@escopecz awesome 👍

**Comment by skripka** [comment]
@escopecz any updates on this? If workload is the issue happy to help out

**Comment by fastyzfr1** [comment]
Wow this issue has been open for almost 5 years! Kudos for you for still working on it. @escopecz any ETA on pushing over the line? It is still in WIP status and was estimated to be out in the 3.1, 3.2 and possibly the 3.3 release but hasn't moved in quite a while.

**Comment by RCheesley** [comment]
@fastyzfr1 there were a *lot* of blocking PRs which needed to get merged before this was able to be considered (many of which have been tested and merged in the 3.x series), and just recently the engineers at Acquia have been debugging some issues with the feature. We are hopeful that it should be able to come over to the community soon once all of that work is complete!

**Comment by fastyzfr1** [comment]
@RCheesley thanks for the update. Little late on my reply so I guess I could ask again. Not sure what soon means in the idea of Mautic and Acquia, since it is now almost 4 months later. Haha

**Comment by npracht** [comment]
Solution : https://github.com/mautic/mautic/pull/12275
Needs to be polished a little bit but working great on our side.
